vendor/k8s.io/klog/v2/README.md

```diff
@@ -23,6 +23,18 @@ Historical context is available here:
  * https://groups.google.com/forum/#!msg/kubernetes-sig-architecture/wCWiWf3Juzs/hXRVBH90CgAJ
  * https://groups.google.com/forum/#!msg/kubernetes-dev/7vnijOMhLS0/1oRiNtigBgAJ
 
+## Release versioning
+
+Semantic versioning is used in this repository. It contains several Go modules
+with different levels of stability:
+- `k8s.io/klog/v2` - stable API, `vX.Y.Z` tags
+- `k8s.io/tools` - no stable API yet (may change eventually), `tools/v0.Y.Z` tags
+- `examples` - no stable API, no tags, no intention to ever stabilize
+
+Exempt from the API stability guarantee are items (packages, functions, etc.)
+which are marked explicitly as `EXPERIMENTAL` in their docs comment. Those
```

Comment:
how will we prevent use of these functions in k8s.io/kubernetes?

Reply:
We *want* to use them in Kubernetes for contextual logging, a new alpha feature.

The rationale for marking them as experimental in klog is that this is our escape hatch from the "v2" API stability guarantee that we offer there: if it turns out that we need to modify them, we still can.

I understand that we need to be extremely careful with this because once we use them in Kubernetes, downstream users can no longer pick an arbitrary klog/v2 and expect it to work. I don't see a good way around that. If we declare them as stable immediately, then we may end up painting ourselves into a corner like we did with `KObjs` which probably should be changed for performance reasons but now has an API that is locked to the current implementation. We also cannot implement these functions outside of klog because they depend on internals that we don't want to export.

Reply:
as soon as we take dependencies on these, and those dependencies propagate out to consumers of k8s client libraries, it becomes pretty unreasonable to break this API in klog/v2...

Reply:
Even when the Kubernetes code itself doesn't offer a stable API? I understand the problem, I'm just trying to determine whether it is a "we must not do this" or a "we should not do this".

I think we can commit to these new API calls as stable already now. I've used them extensively for the contexual logging conversion of kube-scheduler and component-base/logs, therefore I am confident that no further incompatible changes will be needed. It just felt safer to also have a plan B.

Reply:
> Even when the Kubernetes code itself doesn't offer a stable API? I understand the problem, I'm just trying to determine whether it is a "we must not do this" or a "we should not do this".

Yeah, we don't allow k8s.io/kubernetes dependencies to depend back on k8s.io/kubernetes or staging modules, so we avoid gridlock.

Things like k8s.io/utils and k8s.io/klog are fair game for anyone to use though, so it's entirely possible for k8s.io/kubernetes to have diamond dependencies on those modules directly and via transitive deps (we already do).

That's why those projects have to maintain compile compatibility much more strictly (and why we run apidiff on PRs to those projects), to ensure k8s.io/kubernetes is never blocked from picking up an update to k8s.io/utils or k8s.io/klog because a transitive dependency is using a compile-incompatible intermediate version of them.

Reply:
That also rules out introducing experimental new functions even when we then don't use them in Kubernetes: it could happen that dependency A and dependency B of Kubernetes start using them and need different, incompatible releases of klog.

This approach with "EXPERIMENTAL" is from gRPC (for example, https://github.com/grpc/grpc-go/blob/f95b001a48df30dbbdb89ccb9a4781a242591651/reflection/serverreflection.go#L77-L80). gopkg even renders it differently: https://pkg.go.dev/google.golang.org/grpc/reflection#hdr-Experimental

But that must have the same problems. I guess there really is no good way of adding experimental new APIs to a v1 or higher Go package :cry:

Reply:
=> https://github.com/kubernetes/klog/pull/311

Reply:
@liggitt is the proposal in that PR to limit "experimental" to code that is used only for unit tests (like the new ktesting Logger)  acceptable?

Because such code doesn't get pulled into downstream consumers of a package, incompatibilities between different dependencies aren't a problem.

This usage restriction could be enforced by logcheck in a totally generic way: usage of any type or variable that is marked "Experimental" could raise an error except in `_test.go` files. That would also catch usage of things like these experimental gRPC calls.

Reply:
@liggitt please take another look, this PR now uses klog v2.60.1.

We went ahead and removed the "experimental" from the new contextual logging API. The intention (as documented in the revised README.md) is to use "experimental" only for code which is meant to be used in unit tests (https://github.com/kubernetes/klog/tree/main/ktesting, https://github.com/kubernetes/klog/tree/main/test).

"Experimental" is still there for https://github.com/kubernetes/klog/tree/main/textlogger for now but should get removed, because that code might be used by non-test code. The API should be easy to support as stable and it's unlikely that there will be breaking changes (uses functional options, very narrow API for creating the logger).